### ISSUE TITLE
Add hyper-params clustering visualization/selection

### ISSUE THREAD

**Issue by arita37** [optuna/optuna#1706]
## Motivation

 When doing hyper-search, we want to select the ones which are stable and not necessary the most accurate:
    reduce over-fitting risk, and add stability.

 Stability condition is : in the neighborhood of hyper params optimal vector should not be empty.
 Mostly, doing clustering is hyper-params is usefula to pick up the one which has no empty neighborhood.

If your feature request is related to a problem, please describe a clear and concise description of what the problem is. -->

## Description
    Please add clsutering with UMAP to select the hyper params which is most stable + accurate.
    Crieria of seclection   :  
       Minimal; cost value, Maximal number of neighborhood.

Metric :  
   OneHot for category + scalar for num hyper --> euclidian distance.
   UMAP for clsutering, dimension reduction

**Comment by HideakiImamura** [comment]
Thank you for raising an interesting issue. I would like to check on the definition of high stability hyperparameter.

In my understanding, high stability is as follows. For a hyperparameter x, we want to minimize the validation accuracy: the observation of the validation accuracy is usually accompanied by noise. Therefore, the observed value of validation accuracy is stochastic. The hyperparameters with high stability are those with low noise on the validation accuracy.

Is this definition correct?

**Comment by arita37** [comment]
1) how do you measure noise in no-stationnary distribution ?

Think cross val is not enough.


2) My hypothesis is :
   Continuous neighbourhood of optimal hyper params is still optimal (at some epsilon).


3) this is often case than optimal 
hyper params overfits the cross val distribution....

**Comment by arita37** [comment]
Technically,
   Hyper-params optim. is a form of learning over
   (Train —> Val)  distribution.
   It is still subject to “overfitting” over the sampled distribution (assumption is stationarity IID,..)

**Comment by HideakiImamura** [comment]
> 1) how do you measure noise in no-stationnary distribution ?

What is the domain space of that distribution? I assume it as an objective space. I think, for example, the sum of the mean and the variance of the observation `y \sim N(\mu(x), \sigma^2(x))` can be one measurement in the setting, where the distribution of the objective function is not IID. [This paper](https://papers.nips.cc/paper/7818-adversarially-robust-optimization-with-gaussian-processes.pdf) shows how we optimize the non-IID distributed objective function in the Bayesian optimization setting.

By the way, it seems that you think a different setting. Do you want to do some "transfer learning"? The dataset to train the machine learning algorithm has a distribution. You want to find a "good" hyperparameter x, which has a good `E[f(x)]` and a low `Var[f(x)]`, where `f` is an objective function and `E[]` and `Var[]` are taken w.r.t. the distribution of the dataset?

**Comment by arita37** [comment]
Hyper-params is usually done in transition
Distribution of dataset

Let me clarify notations

Xtrain, Xval ,  HyperD are distributions.

When doing Hyper , objective is follow
fit(Xtrain) . eval(Xtrain) : XX Not good
fit(Xtrain) . eval(Xval) 

So, 
   Hyper_fit( Xtrain —> Xval)

Sometimes, this is still overfits,


As you mentionned, there are few ways to tackle it  (need to find relevant paper is any !)

 A) Modified Objective function which include mean, variance of the hyper-params  as additonal term.

     MyLoss + Sum_VarianceHyper * Coef.

    We accept higher Myloss because the variance is small.... 
  
     B) On the final output :
         (loss, hyper_params) ,  we select the hyper_params which are “clustered” within some epsilon level + minimal loss for each cluster....


Obviously, actual paper (or future paper) will be more relevant....
